Automattic/woocommerce-payments
# Fix an issue that led to the payment_type meta not being sent in API requests

Fixes https://github.com/Automattic/woocommerce-payments-server/issues/3619

#### Changes proposed in this Pull Request

In WC Payments https://github.com/Automattic/woocommerce-payments/pull/5585, released as part of 5.6.0, the `Payment_Type` class changed from extending the `MyCLabs\Enum\Enum` type to the `Base_Constant` type. 

This caused the payment type in payment requests to not be sent to server because when it was json encoded [here](https://github.com/Automattic/woocommerce-payments/blob/5.6.0/includes/wc-payment-api/class-wc-payments-api-client.php#L2064), it would be stripped out. 

eg 

```
[metadata] => Array
(
    [customer_name] => James Allan
    [customer_email] => james.allan@automattic.com
    [site_url] => https://wcdev.local
    [order_id] => 548
    [order_number] => 548
    [order_key] => wc_order_K0lx0Zre7aglL
    [payment_type] => WCPay\Constants\Payment_Type Object
        (
            [value:protected] => RECURRING
        )
    [fraud_prevention_data_shopper_ip_hash] => 
    [fraud_prevention_data_ip_country] => AU
    [fraud_prevention_data_available] => 1
)
```

When encoded was being turned into 

```
[metadata] => stdClass Object
(
    [customer_name] => James Allan
    [customer_email] => james.allan@automattic.com
    [site_url] => https://wcdev.local
    [order_id] => 551
    [order_number] => 551
    [order_key] => wc_order_Zd13ZFxnWGTAT
    [payment_type] => stdClass Object
        (
        )
    [fraud_prevention_data_shopper_ip_hash] => 
    [fraud_prevention_data_ip_country] => AU
    [fraud_prevention_data_available] => 1
)
```

This PR fixes that by implementing the `jsonSerialize` function. 

#### Testing instructions





1. Enable the WC Subscriptions plugin.
2. Create a [simple subscription product](https://woocommerce.com/document/subscriptions/store-manager-guide/#section-1). 
3. Purchase the subscription. 
4. In the WP admin go to **WooCommerce → Subscriptions** and edit the subscription. 
5. From the [actions dropdown](https://woocommerce.com/document/subscriptions/add-or-modify-a-subscription/#section-15), choose the option to **"Process renewal"** 
6. Save. 
7. Go to your Stripe dashboard and look at the last transaction. 
8. Scroll down to the **Metadata** section.
   - On `develop` you'll notice there is no `payment_type` arg. 
   - On this branch you should see the `'recurring'` payment type.


| `develop` | This branch |
|--------|--------|
| <img width="899" alt="Screenshot 2023-07-13 at 12 06 49 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/f8e222ee-b461-4f84-a128-592766f707b3"> | <img width="913" alt="Screenshot 2023-07-13 at 12 04 36 pm" src="https://github.com/Automattic/woocommerce-payments/assets/8490476/af2fd590-4574-427f-ba0d-f53c4cb61e30"> | 


If you want to look at the raw code that's in question, observe the `$params` value [here](https://github.com/Automattic/woocommerce-payments/blob/5.9.0/includes/wc-payment-api/class-wc-payments-api-client.php#L1753) and the `$body` result. You'll notice the payment type is being removed. 

-------------------


**Post merge**
